lexer.py

```diff
@@ -0,0 +1,142 @@
+from tokens import TokenTypes, Token
+
+###############################################################################
+#                                                                             #
+#   LEXER                                                                     #
+#                                                                             #
+###############################################################################
+
+ESCCHARS = {
+    "n":  "\n",
+    "t":  "\t",
+    "\\": "\\",
+    "'": "'",
+    "r":  "\r",
+}
+
+RESERVED_KEYWORDS = {
+            "SELECT": lambda x: Token(TokenTypes.SELECT, "SELECT", x),
+            "FROM": lambda x: Token(TokenTypes.FROM, "FROM", x),
+            "EDGES": lambda x: Token(TokenTypes.EDGES, "EDGES", x),
+            "NODES": lambda x: Token(TokenTypes.NODES, "NODES", x),
+            "TRAVERSE": lambda x: Token(TokenTypes.TRAVERSE, "TRAVERSE", x),
+            "BY": lambda x: Token(TokenTypes.BY, "BY", x),
+            "DEPTH": lambda x: Token(TokenTypes.DEPTH, "DEPTH", x),
+            "BREADTH": lambda x: Token(TokenTypes.BREADTH, "BREADTH", x),
+            "WHERE": lambda x: Token(TokenTypes.WHERE, "WHERE", x),
+            "LIMIT": lambda x: Token(TokenTypes.LIMIT, "LIMIT", x),
+            "TRUE": lambda x: Token(TokenTypes.TRUE, True, x),
+            "FALSE": lambda x: Token(TokenTypes.FALSE, False, x),
+            "AND": lambda x: Token(TokenTypes.AND, "AND", x),
+            "OR": lambda x: Token(TokenTypes.OR, "OR", x),
+            "NOT": lambda x: Token(TokenTypes.NOT, "NOT", x),
+}
+
+class Lexer(object):
+    def __init__(self, text):
+        self.text = text         # string input
+        self.pos = 0             # self.pos is index into self.text
+        self.line = 1            # current line
+        self.current_char = self.text[0] # current char
+        self.tokens = []         # tokens so far
+    def error(self, msg=""):
+        raise Exception("LexingError, line %s: " + msg + "." % self.line)
+    def peek(self, lookahead=1):
+        peek_pos = self.pos + lookahead
+        if peek_pos >= len(self.text):
+            return None
+        else:
+            return self.text[peek_pos]
+    def advance(self):
```

Comment:
Should `error`, `peek`, and `advance` be private?

Reply:
If they should, consider making `Lexer` a function instead of a class, which will help with all the `self.` noise.